File: app/src/main/java/com/foxxyz/chronovision/FileUploadService.java
package com.foxxyz.chronovision;

import android.app.IntentService;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.ResultReceiver;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class FileUploadService extends IntentService {
    public static final int RUNNING = 1;
    public static final int ERROR = 2;
    public static final int FINISHED = 3;

    public FileUploadService() {
        super("FileUploadService");
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        String url = intent.getStringExtra("url");
        OkHttpClient.Builder builder = new OkHttpClient.Builder();
        builder.connectTimeout(30, TimeUnit.SECONDS);
        builder.readTimeout(30, TimeUnit.SECONDS);
        Uri photo = intent.getParcelableExtra("photo");

        // Set receiver
        final ResultReceiver receiver = intent.getParcelableExtra("receiver");

        // Inform of progress
        Bundle b = new Bundle();
        receiver.send(RUNNING, Bundle.EMPTY);

        // Start the request
        try {
            MultipartBody.Builder body = new MultipartBody.Builder()
                    .setType(MultipartBody.FORM)
                    .addFormDataPart("date", intent.getStringExtra("date"))
                    .addFormDataPart("photo", "photo.jpg", new InputStreamRequestBody(MediaType.parse("image/jpg"), getContentResolver(), photo));
            Request.Builder request = new Request.Builder().url(url);
            Response response = post(builder, request, body);
            receiver.send(FINISHED, Bundle.EMPTY);
        }
        catch (IOException e) {
            e.printStackTrace();
            b.putString(Intent.EXTRA_TEXT, e.toString());
            receiver.send(ERROR, b);
        }
    }

    public Response post(OkHttpClient.Builder builder, Request.Builder requestBuilder, MultipartBody.Builder bodyBuilder) throws IOException {
        OkHttpClient client = builder.build();
        Request request = requestBuilder.post(bodyBuilder.build()).build();
        Response response = client.newCall(request).execute();
        if (!response.isSuccessful()) throw new IOException("Unexpected code " + response);
        return response;
    }
}

File: app/src/main/java/com/foxxyz/chronovision/FileUploadService.kt
package com.foxxyz.chronovision

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.ResultReceiver
import androidx.work.Worker

import java.io.IOException
import java.util.concurrent.TimeUnit

import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response

class FileUploadService : Worker {

    override fun onHandleIntent(intent: Intent?) {
        val url = intent!!.getStringExtra("url")
        val authToken = intent.getStringExtra("authToken")
        val builder = OkHttpClient.Builder()
        builder.connectTimeout(10, TimeUnit.SECONDS)
        builder.readTimeout(10, TimeUnit.SECONDS)
        val photo = intent.getParcelableExtra<Uri>("photo")

        // Set receiver
        val receiver = intent.getParcelableExtra<ResultReceiver>("receiver")

        // Inform of progress
        val b = Bundle()
        receiver?.send(RUNNING, Bundle.EMPTY)

        // Start the request
        try {
            val mType = "image/jpg".toMediaTypeOrNull()!!
            val body = intent.getStringExtra("date")?.let {
                MultipartBody.Builder()
                        .setType(MultipartBody.FORM)
                        .addFormDataPart("date", it)
                        .addFormDataPart("photo", "photo.jpg", InputStreamRequestBody(mType, contentResolver, photo))
            }
            val request = url?.let {
                Request.Builder().url(it)
                        .addHeader("Authorization", "Bearer $authToken")
            }
            if (request != null && body != null) {
                Companion.post(builder, request, body)
            }
            receiver?.send(FINISHED, Bundle.EMPTY)
        } catch (e: IOException) {
            e.printStackTrace()
            b.putString(Intent.EXTRA_TEXT, e.toString())
            receiver?.send(ERROR, b)
        }

    }

    companion object {
        const val RUNNING = 1
        const val ERROR = 2
        const val FINISHED = 3

        @Throws(IOException::class)
        fun post(builder: OkHttpClient.Builder, requestBuilder: Request.Builder, bodyBuilder: MultipartBody.Builder): Response {
            val client = builder.build()
            val request = requestBuilder.post(bodyBuilder.build()).build()
            val response = client.newCall(request).execute()
            if (!response.isSuccessful) throw IOException("Unexpected code $response")
            return response
        }
    }
}
